File: filter/filter_test.go
package filter

import (
	"testing"
)

func TestStringFilter_Validate(t *testing.T) {
	tests := []struct {
		name    string
		filter  *StringFilter
		wantErr bool
	}{
		{
			name:    "nil filter",
			filter:  nil,
			wantErr: false,
		},
		{
			name:    "empty filter",
			filter:  &StringFilter{},
			wantErr: false,
		},
		{
			name: "valid regex",
			filter: &StringFilter{
				Re: "^test.*",
			},
			wantErr: false,
		},
		{
			name: "valid complex regex",
			filter: &StringFilter{
				Re: "\\d+",
			},
			wantErr: false,
		},
		{
			name: "invalid regex",
			filter: &StringFilter{
				Re: "[",
			},
			wantErr: true,
		},
		{
			name: "invalid regex with escape",
			filter: &StringFilter{
				Re: "\\",
			},
			wantErr: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			err := tt.filter.Validate()
			if (err != nil) != tt.wantErr {
				t.Fatalf("StringFilter.Validate() error = %v, wantErr %v", err, tt.wantErr)
			}
		})
	}
}

func TestStringFilter_CheckMatch(t *testing.T) {
	tests := []struct {
		name   string
		filter *StringFilter
		value  string
		want   bool
	}{
		{
			name:   "nil filter",
			filter: nil,
			value:  "test",
			want:   true,
		},
		{
			name:   "empty filter",
			filter: &StringFilter{},
			value:  "test",
			want:   true,
		},
		{
			name: "empty=true matches empty string",
			filter: &StringFilter{
				Empty: true,
			},
			value: "",
			want:  true,
		},
		{
			name: "empty=true does not match non-empty string",
			filter: &StringFilter{
				Empty: true,
			},
			value: "test",
			want:  false,
		},
		{
			name: "notEmpty=true matches non-empty string",
			filter: &StringFilter{
				NotEmpty: true,
			},
			value: "test",
			want:  true,
		},
		{
			name: "notEmpty=true does not match empty string",
			filter: &StringFilter{
				NotEmpty: true,
			},
			value: "",
			want:  false,
		},
		{
			name: "value matches exact value",
			filter: &StringFilter{
				Value: "test",
			},
			value: "test",
			want:  true,
		},
		{
			name: "value does not match different value",
			filter: &StringFilter{
				Value: "test",
			},
			value: "other",
			want:  false,
		},
		{
			name: "values matches when value in array",
			filter: &StringFilter{
				Values: []string{"test1", "test2"},
			},
			value: "test1",
			want:  true,
		},
		{
			name: "values does not match when value not in array",
			filter: &StringFilter{
				Values: []string{"test1", "test2"},
			},
			value: "test3",
			want:  false,
		},
		{
			name: "empty values array matches any value",
			filter: &StringFilter{
				Values: []string{},
			},
			value: "test",
			want:  true,
		},
		{
			name: "regex matches",
			filter: &StringFilter{
				Re: "^test",
			},
			value: "test123",
			want:  true,
		},
		{
			name: "regex does not match",
			filter: &StringFilter{
				Re: "^test",
			},
			value: "abc123",
			want:  false,
		},
		{
			name: "invalid regex returns false",
			filter: &StringFilter{
				Re: "[",
			},
			value: "test",
			want:  false,
		},
		{
			name: "hasPrefix matches",
			filter: &StringFilter{
				HasPrefix: "test",
			},
			value: "test123",
			want:  true,
		},
		{
			name: "hasPrefix does not match",
			filter: &StringFilter{
				HasPrefix: "test",
			},
			value: "abc123",
			want:  false,
		},
		{
			name: "hasSuffix matches",
			filter: &StringFilter{
				HasSuffix: "123",
			},
			value: "test123",
			want:  true,
		},
		{
			name: "hasSuffix does not match",
			filter: &StringFilter{
				HasSuffix: "123",
			},
			value: "test456",
			want:  false,
		},
		{
			name: "contains matches",
			filter: &StringFilter{
				Contains: "est",
			},
			value: "test123",
			want:  true,
		},
		{
			name: "contains does not match",
			filter: &StringFilter{
				Contains: "xyz",
			},
			value: "test123",
			want:  false,
		},
		{
			name: "combined filters all match",
			filter: &StringFilter{
				NotEmpty:  true,
				HasPrefix: "test",
				HasSuffix: "123",
			},
			value: "test123",
			want:  true,
		},
		{
			name: "combined filters one fails",
			filter: &StringFilter{
				NotEmpty:  true,
				HasPrefix: "test",
				HasSuffix: "456",
			},
			value: "test123",
			want:  false,
		},
		{
			name: "complex combination",
			filter: &StringFilter{
				Re:        "^test",
				Contains:  "12",
				HasSuffix: "3",
			},
			value: "test123",
			want:  true,
		},
		{
			name: "regex with digits",
			filter: &StringFilter{
				Re: "\\d+",
			},
			value: "abc123",
			want:  true,
		},
		{
			name: "regex with digits no match",
			filter: &StringFilter{
				Re: "\\d+",
			},
			value: "abcdef",
			want:  false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := tt.filter.CheckMatch(tt.value)
			if got != tt.want {
				t.Fatalf("StringFilter.CheckMatch() = %v, want %v", got, tt.want)
			}
		})
	}
}

File: filter/filter.test.ts
import { describe, it, expect } from 'vitest'
import { validateStringFilter, checkStringFilterMatch } from './filter.js'
import { StringFilter } from './filter.pb.js'

describe('validateStringFilter', () => {
  it('should return null for null/undefined filter', () => {
    expect(validateStringFilter(null)).toBeNull()
    expect(validateStringFilter(undefined)).toBeNull()
  })

  it('should return null for empty filter', () => {
    expect(validateStringFilter({})).toBeNull()
  })

  it('should return null for valid regex', () => {
    expect(validateStringFilter({ re: '^test.*' })).toBeNull()
    expect(validateStringFilter({ re: '[0-9]+' })).toBeNull()
  })

  it('should return error message for invalid regex', () => {
    const result = validateStringFilter({ re: '[' })
    expect(result).toContain('re:')
    expect(result).toContain('Invalid regular expression')
  })

  it('should handle regex with escape sequences', () => {
    expect(validateStringFilter({ re: '\\d+' })).toBeNull()
  })
})

describe('checkStringFilterMatch', () => {
  it('should return true for null/undefined filter', () => {
    expect(checkStringFilterMatch(null, 'test')).toBe(true)
    expect(checkStringFilterMatch(undefined, 'test')).toBe(true)
  })

  it('should return true for empty filter', () => {
    expect(checkStringFilterMatch({}, 'test')).toBe(true)
  })

  describe('empty filter', () => {
    it('should match empty string when empty=true', () => {
      expect(checkStringFilterMatch({ empty: true }, '')).toBe(true)
    })

    it('should not match non-empty string when empty=true', () => {
      expect(checkStringFilterMatch({ empty: true }, 'test')).toBe(false)
    })
  })

  describe('notEmpty filter', () => {
    it('should match non-empty string when notEmpty=true', () => {
      expect(checkStringFilterMatch({ notEmpty: true }, 'test')).toBe(true)
    })

    it('should not match empty string when notEmpty=true', () => {
      expect(checkStringFilterMatch({ notEmpty: true }, '')).toBe(false)
    })
  })

  describe('value filter', () => {
    it('should match exact value', () => {
      expect(checkStringFilterMatch({ value: 'test' }, 'test')).toBe(true)
    })

    it('should not match different value', () => {
      expect(checkStringFilterMatch({ value: 'test' }, 'other')).toBe(false)
    })
  })

  describe('values filter', () => {
    it('should match when value is in values array', () => {
      expect(checkStringFilterMatch({ values: ['test1', 'test2'] }, 'test1')).toBe(true)
      expect(checkStringFilterMatch({ values: ['test1', 'test2'] }, 'test2')).toBe(true)
    })

    it('should not match when value is not in values array', () => {
      expect(checkStringFilterMatch({ values: ['test1', 'test2'] }, 'test3')).toBe(false)
    })

    it('should return true when values array is empty', () => {
      expect(checkStringFilterMatch({ values: [] }, 'test')).toBe(true)
    })
  })

  describe('regex filter', () => {
    it('should match when regex matches', () => {
      expect(checkStringFilterMatch({ re: '^test' }, 'test123')).toBe(true)
      expect(checkStringFilterMatch({ re: '\\d+' }, 'abc123')).toBe(true)
    })

    it('should not match when regex does not match', () => {
      expect(checkStringFilterMatch({ re: '^test' }, 'abc123')).toBe(false)
      expect(checkStringFilterMatch({ re: '\\d+' }, 'abcdef')).toBe(false)
    })

    it('should return false for invalid regex', () => {
      expect(checkStringFilterMatch({ re: '[' }, 'test')).toBe(false)
    })
  })

  describe('hasPrefix filter', () => {
    it('should match when string has prefix', () => {
      expect(checkStringFilterMatch({ hasPrefix: 'test' }, 'test123')).toBe(true)
    })

    it('should not match when string does not have prefix', () => {
      expect(checkStringFilterMatch({ hasPrefix: 'test' }, 'abc123')).toBe(false)
    })
  })

  describe('hasSuffix filter', () => {
    it('should match when string has suffix', () => {
      expect(checkStringFilterMatch({ hasSuffix: '123' }, 'test123')).toBe(true)
    })

    it('should not match when string does not have suffix', () => {
      expect(checkStringFilterMatch({ hasSuffix: '123' }, 'test456')).toBe(false)
    })
  })

  describe('contains filter', () => {
    it('should match when string contains substring', () => {
      expect(checkStringFilterMatch({ contains: 'est' }, 'test123')).toBe(true)
    })

    it('should not match when string does not contain substring', () => {
      expect(checkStringFilterMatch({ contains: 'xyz' }, 'test123')).toBe(false)
    })
  })

  describe('combined filters', () => {
    it('should match when all filters match', () => {
      expect(checkStringFilterMatch({ 
        notEmpty: true, 
        hasPrefix: 'test', 
        hasSuffix: '123' 
      }, 'test123')).toBe(true)
    })

    it('should not match when any filter fails', () => {
      expect(checkStringFilterMatch({ 
        notEmpty: true, 
        hasPrefix: 'test', 
        hasSuffix: '456' 
      }, 'test123')).toBe(false)
    })

    it('should handle complex combination', () => {
      expect(checkStringFilterMatch({ 
        re: '^test', 
        contains: '12', 
        hasSuffix: '3' 
      }, 'test123')).toBe(true)
    })
  })
})
